add more info to `project` response

enricosada opened the pull request:
Added new attributes to `project` response (so is just an addition, not breaking). 

- `SdkType`  the type of the project (`dotnet/sdk` for new sdk, `verbose` for old sdk, `project.json` for well, project.json)
- `OutputType` if `lib` or `exe` or others
- `AdditionalInfo` a map of additional props, to be used later.

idea is to return some useful (free, so no performance cost) info from project parsing, who can be useful for consumer

- the `SdkType` may be nice to set the default msbuild host runtime (what use for `Build`? msbuild or .net cli? just check `SdkType` for a sane default). Or enable new sdk commands.
- `OutputType` can be used to filter the starting project maybe, or anyway if current can be run or not.



The `AdditionalInfo` something i'd like to set as base, for extract more info to help consumer to enable/disable integrations or commands.
atm is a map. i'd like to add more info (who are dynamic per project usually).
For example:

- `IsTestProject`? `TestFramework` as `nunit/expecto/etc`? to enable `dotnet test` or additional extension like for expecto
- `IsPackable`? this is a new property of .net sdk, can be good to filter msbuild commands
- `Fable`? `Paket`?

Pratically move all the usual checks, from consumer to FSAC, because are easier to do and shared.

Maybe a better name is `Capabilities`.
So array `AdditionaInfo` become `Capabilities: Capability array` and

```fsharp
type Capability = { Name: string; Data: Map<string, string> }
```

@rneatherway @kjnilsson @Krzysztof-Cieslak there are info (per project) you need?

TODO:

- [x] ~use option json converter in command line based?~ no. used by linter api (so a breaking change)
- [x] make `RunCmd` typed not tuple

enricosada committed:
refactoring json write only converters to functions

enricosada committed:
add SdkType, OutputType, AdditionalInfo to project response

- SdkType: the type of the project
- OutputType: if lib or exe or others
- AdditionalInfo: a map of additional props

enricosada committed:
refactor to not duplicated json converters

Krzysztof-Cieslak commented on src/FsAutoComplete.Core/Commands.fs:
```diff
@@ -137,20 +137,26 @@ type Commands (serialize : Serializer) =
+                        | Some "exe" -> ProjectOutputType.Exe
+                        | Some "library" -> ProjectOutputType.Library
+                        | Some x -> ProjectOutputType.Custom x
+                        | None -> ProjectOutputType.Exe //the default
```
Is defaulting to `exe` for sure safe choice?

enricosada replied:
got catch, was wrong, set it to `library` as default

Krzysztof-Cieslak commented:
Ionide at the moment needs:
1. `SdkType` for `MsBuild / Expecto` integrations
2. `OutputType` for Expecto integration (and better handling of build / debug configurations)
3. `IsTests` for `Expecto integration` (right now we just check if `Expecto.dll` is in the references list for each project.

enricosada committed:
use ExtraProjectInfo to store additional project data

like ProjectSdkType and ProjectOutputType

for new sdk (dotnet/sdk) more info are avaiable

enricosada committed:
refactor extract f

enricosada commented:
ok made response more typed and based on sdk type.

```diff
  type ProjectResponse =
    {
      Project: ProjectFilePath
      Files: List<SourceFilePath>
      Output: string
      References: List<ProjectFilePath>
      Logs: Map<string, string>
+     OutputType: ProjectOutputType
+     Info: ProjectResponseInfo
+     AdditionalInfo: Map<string, string>
    }
```

and info are

```fsharp
  type ProjectResponseInfo =
    | DotnetSdk of ProjectResponseInfoDotnetSdk
    | Verbose // old fsproj
    | ProjectJson
  and ProjectResponseInfoDotnetSdk =
    {
      IsTestProject: bool
      Configuration: string
      IsPackable: bool
      TargetFramework: string
      TargetFrameworkIdentifier: string
      TargetFrameworkVersion: string
      MSBuildRuntimeType: string
      RestoreSuccess: bool
      Configurations: string list
      TargetFrameworks: string list
      RunCmd: (string * string) option // cmd and args, same of `dotnet run` (.net core) or .exe (.net)
      IsPublishable: bool option // avaiable from sdk2.0 only
      BundledNETCoreAppTargetFrameworkVersion: string option // avaiable from sdk2.0 only
      BundledNETStandardTargetFrameworkVersion: string option // avaiable from sdk2.0 only
    }
```

the info are stored in the `FSharpProjectOption.ExtraProjectInfo` (before unused = None) so always avaiable and doesnt impact a lot.

```fsharp
type ExtraProjectInfoData =
    {
        ProjectOutputType: ProjectOutputType
        ProjectSdkType: ProjectSdkType
    }
and ProjectOutputType =
    | Library
    | Exe
    | Custom of string

```

About `AdditionalInfo`, these are for misc info, like `IsFable`, `IsExcepto` etc, who are just inferred by an algorithm, not read from msbuild project directly.
For example atm we can check if `IsExpecto` if the dll exist and `IsTestProject`, etc.

enricosada committed:
refactor, extract project parsing type in a file

enricosada committed:
declaring bankruptcy on old fsproj and project.json

response will return empty object, so props can be added later, if at all
will use additionalinfo to return inferred info, like istest or expecto

enricosada committed:
option was never used in public api

enricosada committed:
removed runtime props (not specific of project itself, but of execution of msbuild)

enricosada committed:
tuple is serialized ugly

enricosada committed:
Revert "option was never used in public api"

option are used by linter, avoid the breaking change

This reverts commit 385995ce40c23bae1f7fb84e85f0c16e70d98cf3.

enricosada committed:
fix output type for old fsproj, project.json

enricosada committed:
updated tests

rneatherway commented:
It's all good for me, except the tests seem not to be consistent between mono and windows.

Krzysztof-Cieslak commented:
I'm not sure if I'm fan of exposing DUs in output types - serialization / deserialization of them will be annoying, default JSON.Net serialization is rather ugly

enricosada commented:
@Krzysztof-Cieslak yes, seen that. this is readonly, so easier to just serialize.
i added a custom converter for that in https://github.com/fsharp/FsAutoComplete/pull/187/files#diff-3228934b0a597261234f0059ebc0e18dR30

```fsharp
  type ProjectResponseInfo =
    | DotnetSdk of ProjectResponseInfoDotnetSdk
    | Verbose
    | ProjectJson
  and ProjectResponseInfoDotnetSdk =
    {
      IsTestProject: bool
      Configuration: string
      IsPackable: bool
      TargetFramework: string
      TargetFrameworkIdentifier: string
      TargetFrameworkVersion: string
      RestoreSuccess: bool
      Configurations: string list
      TargetFrameworks: string list
      RunCmd: RunCmd option
      IsPublishable: bool option
    }
  and [<RequireQualifiedAccess>] RunCmd = { Command: string; Arguments: string }
```

is serialized as usual case + data attributes

```json
  {
      "SdkType": "dotnet/sdk",
      "Data": {
          ,,,
      }
  }
```

sample output:

```json
    "Info": {
      "SdkType": "dotnet/sdk",
      "Data": {
        "IsTestProject": false,
        "Configuration": "Debug",
        "IsPackable": true,
        "TargetFramework": "netcoreapp1.0",
        "TargetFrameworkIdentifier": ".NETCoreApp",
        "TargetFrameworkVersion": "v1.0",
        "RestoreSuccess": true,
        "Configurations": [
          "Debug",
          "Release"
        ],
        "TargetFrameworks": [],
        "RunCmd": {
              "Command": "dotnet",
              "Arguments": "exec \"<absolute path removed>/FsAutoComplete.IntegrationTests/DotNetCore/sample1/c1/bin/Debug/netcoreapp1.0/c1.dll\""
            }
        },
        "IsPublishable": null
      }
    },
```

alternative? the only thing i can think about, is to specify the types so these include the case attribute (like `"SdkType": "dotnet/sdk"` inside), but anyway consumer need an if over the DU case field because the properties are different.
I am ok with everything, it doesnt matter a lot

enricosada committed:
removed Configurations because is not avaiable on unix/osx

enricosada commented:
@rneatherway yes, seems different. i removed the property for now, i'll check why it differ (it shouldnt)

Krzysztof-Cieslak commented:
Looking at the output format in tests, it seems OK for me. The only small problem is `option` serialization but it can be handled later.

enricosada commented:
> The only small problem is option serialization but it can be handled later.

That's just for `stdio`, the `suave` use an option converter for serialziation (None -> Null, Some x -> x).

But i didnt want to apply that converter on stdio too because is a breaking change (option are used by Linter), so can be done later maybe if both @rneatherway and @kjnilsson agree to consolidate it.

Krzysztof-Cieslak commented:
TBF, I'm not sure if either Emacs or vim plugins are using Linter at all. It may be the reason why this bad serialization still exists

enricosada commented:
@rneatherway this is ok for final review and merge ihmo.
Add just more attribute to `parse` response (so is not a breaking change), if you want to consolitate serializers (fixing the ugly Option, like suave does), i can do it in another PR

rneatherway commented:
Yes thanks, looks good to me. Emacs does use the linter, you can see the parsing of the message here: https://github.com/fsharp/emacs-fsharp-mode/blob/master/flycheck-fsharp.el#L68

From the look of that code there is no parsing of an option type. I would definitely support your plan to change the serializer to match that for suave frontend.

Krzysztof-Cieslak commented:
>  Emacs does use the linter

Oh, great! Didn't know that :)
